### PR TITLE
ICU-22104 Add tests for CPP alphabetic index set/get maximum label count

### PR DESCRIPTION
**Pull request by sven-oly**
##### Checklist

- [x] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22104
- [x] Required: The PR title must be prefixed with a JIRA Issue number. 
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [x] Issue accepted (done by Technical Committee after discussion)
- [x] Tests included, if applicable
- [x] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

File: icu4c/source/test/intltest/alphaindextst.cpp [hunk 1]
```diff
@@ -756,4 +757,52 @@ void AlphabeticIndexTest::checkHasBuckets(const Locale &locale, UScriptCode scri
...
+    AlphabeticIndex aindex(locale, status);
+
+    int32_t defaultMax = aindex.getMaxLabelCount();
+    assertEquals("Expect max count = 99", 99, defaultMax);
+
+    aindex.setMaxLabelCount(237, status);
+    if (U_FAILURE(errorCode)) {
+      dataerrln("%s %d  Error in index setMaxLabelCount",  __FILE__, __LINE__);
+      return;
+    }
```
Comment (markusicu) [inline icu4c/source/test/intltest/alphaindextst.cpp:774]: Why do you use assertEquals() above but not here?

If you do want it like this, then use errln() here not dataerrln().

dataerrln() is appropriate when a failure at that point might have to do with data loading.
Please check the `status` right after the constructor, and use the `UBool errDataIfFailureAndReset(const char *fmt, ...);` method on the errorCode.

File: icu4c/source/test/intltest/alphaindextst.cpp [hunk 2]
```diff
@@ -756,4 +757,52 @@ void AlphabeticIndexTest::checkHasBuckets(const Locale &locale, UScriptCode scri
             uscript_getScript(bucket->getLabel().char32At(0), errorCode));
 }
 
+void AlphabeticIndexTest::checkMaxLabelCount() {
+    IcuTestErrorCode errorCode(*this, "checkMaxLabelCount");
+    UErrorCode status = U_ZERO_ERROR;
```
Comment (markusicu) [inline icu4c/source/test/intltest/alphaindextst.cpp:762]: Why do you have both `errorCode` and `status`? Just use the higher-function `errorCode`.

File: icu4c/source/test/intltest/alphaindextst.cpp [hunk 3]
```diff
@@ -756,4 +757,52 @@ void AlphabeticIndexTest::checkHasBuckets(const Locale &locale, UScriptCode scri
...
+    int32_t defaultMax = aindex.getMaxLabelCount();
+    assertEquals("Expect max count = 99", 99, defaultMax);
+
+    aindex.setMaxLabelCount(237, status);
+    if (U_FAILURE(errorCode)) {
+      dataerrln("%s %d  Error in index setMaxLabelCount",  __FILE__, __LINE__);
+      return;
+    }
+
+    // Verify that all bucket are cleared after setting max bucket count.
```
Comment (markusicu) [inline icu4c/source/test/intltest/alphaindextst.cpp:776]: All buckets? The code checks only the first one.

File: icu4c/source/test/intltest/alphaindextst.cpp [hunk 4]
```diff
@@ -756,4 +757,52 @@ void AlphabeticIndexTest::checkHasBuckets(const Locale &locale, UScriptCode scri
...
+    const UnicodeString bucketLabel = aindex.getBucketLabel();
+    assertEquals("Expect empty string", "", bucketLabel);
+
+    int32_t maxChanged = aindex.getMaxLabelCount();
+    assertEquals("Expect max count = 237", 237, maxChanged);
+
+    int32_t bigMax = 100000;
+    aindex.setMaxLabelCount(bigMax, errorCode);
```
Comment (markusicu) [inline icu4c/source/test/intltest/alphaindextst.cpp:784]: I don't know that it's useful to set it more than once without doing anything interesting in between.

I suggest you
- set it once to a value a bit higher than 1000
- add almost that many labels
- buildImmutableIndex() and check that you got all of the expected buckets
- add a bunch more labels, so that you exceed the max
- buildImmutableIndex() and check that the number of buckets does not exceed the max, and that the buckets you get are a  subset of those you set

File: icu4c/source/test/intltest/alphaindextst.cpp [hunk 5]
```diff
@@ -756,4 +757,52 @@ void AlphabeticIndexTest::checkHasBuckets(const Locale &locale, UScriptCode scri
...
+    Locale locale("en");
+    UnicodeString loc = locale.getName();
+    AlphabeticIndex aindex(locale, status);
+
+    int32_t defaultMax = aindex.getMaxLabelCount();
+    assertEquals("Expect max count = 99", 99, defaultMax);
+
+    aindex.setMaxLabelCount(237, status);
+    if (U_FAILURE(errorCode)) {
+      dataerrln("%s %d  Error in index setMaxLabelCount",  __FILE__, __LINE__);
```
Comment (markusicu) [inline icu4c/source/test/intltest/alphaindextst.cpp:772]: We always indent by multiples of 4.

File: icu4c/source/test/intltest/alphaindextst.cpp [hunk 6]
```diff
@@ -756,4 +757,52 @@ void AlphabeticIndexTest::checkHasBuckets(const Locale &locale, UScriptCode scri
...
+      return;
+    }
+
+    // Verify that all bucket are cleared after setting max bucket count.
+    const UnicodeString bucketLabel = aindex.getBucketLabel();
```
Comment (markusicu) [inline icu4c/source/test/intltest/alphaindextst.cpp:777]: The function returns a reference.

File: icu4c/source/test/intltest/alphaindextst.cpp [hunk 7]
```diff
@@ -756,4 +757,52 @@ void AlphabeticIndexTest::checkHasBuckets(const Locale &locale, UScriptCode scri
...
+      dataerrln("%s %d  Error in index setMaxLabelCount to bigMax",
+                __FILE__, __LINE__);
+      return;
+    }
+    maxChanged = aindex.getMaxLabelCount();
+    assertEquals("Expect max count = 100000", bigMax, maxChanged);
+
+    int32_t numBuckets = aindex.getBucketCount(status);
+    int32_t expectedBuckets = 0x1C;
+    assertEquals("Expect numBuckets = 1", expectedBuckets, numBuckets);
```
Comment (markusicu) [inline icu4c/source/test/intltest/alphaindextst.cpp:795]: You are not actually expecting it to be 1. I suggest you don't hardcode the expected value in the message, since they are easily out of sync.

File: icu4c/source/test/intltest/alphaindextst.cpp [hunk 8]
```diff
@@ -756,4 +757,52 @@ void AlphabeticIndexTest::checkHasBuckets(const Locale &locale, UScriptCode scri
...
 
+void AlphabeticIndexTest::checkMaxLabelCount() {
+    IcuTestErrorCode errorCode(*this, "checkMaxLabelCount");
+    UErrorCode status = U_ZERO_ERROR;
+    Locale locale("en");
+    UnicodeString loc = locale.getName();
+    AlphabeticIndex aindex(locale, status);
+
+    int32_t defaultMax = aindex.getMaxLabelCount();
+    assertEquals("Expect max count = 99", 99, defaultMax);
```
Comment (markusicu) [inline icu4c/source/test/intltest/alphaindextst.cpp:768]: I would not test that it's a specific number (which is not guaranteed to not change), but that it's, say, fewer than 200.

File: icu4c/source/test/intltest/alphaindextst.cpp [hunk 9]
```diff
@@ -756,4 +757,52 @@ void AlphabeticIndexTest::checkHasBuckets(const Locale &locale, UScriptCode scri
...
+    if (U_FAILURE(errorCode)) {
+      dataerrln("%s %d  Error in index setMaxLabelCount to bigMax",
+                __FILE__, __LINE__);
+      return;
+    }
+    maxChanged = aindex.getMaxLabelCount();
+    assertEquals("Expect max count = 100000", bigMax, maxChanged);
+
+    int32_t numBuckets = aindex.getBucketCount(status);
+    int32_t expectedBuckets = 0x1C;
```
Comment (markusicu) [inline icu4c/source/test/intltest/alphaindextst.cpp:794]: This magic number deserves a comment, and probably wants to be in decimal.

File: icu4c/source/test/intltest/alphaindextst.cpp [hunk 10]
```diff
@@ -756,4 +757,52 @@ void AlphabeticIndexTest::checkHasBuckets(const Locale &locale, UScriptCode scri
             uscript_getScript(bucket->getLabel().char32At(0), errorCode));
 }
 
+void AlphabeticIndexTest::checkMaxLabelCount() {
+    IcuTestErrorCode errorCode(*this, "checkMaxLabelCount");
+    UErrorCode status = U_ZERO_ERROR;
+    Locale locale("en");
+    UnicodeString loc = locale.getName();
```
Comment (markusicu) [inline icu4c/source/test/intltest/alphaindextst.cpp:764]: unused?

File: icu4c/source/test/intltest/alphaindextst.cpp (hunk 7 above)
Comment (aheninger) [reply icu4c/source/test/intltest/alphaindextst.cpp:795]: I'm fond of using just the test framework macro WHERE for expect messages. It gives the file name and line number, which makes it much easier to zero in on a failing test, compared with searching for the text of the error message in  the test source code. Plus it's quick and easy to write; it's all you really need.
